GAE/test/org/akvo/flow/api/app/DataPointServletTest.java

```diff
@@ -0,0 +1,214 @@
+        final SurveyedLocaleDao dpDao = new SurveyedLocaleDao();
+        final List<SurveyedLocale> datapoints = new ArrayList<>();
+        for (int i = 0; i < howMany; i++) {
+            SurveyedLocale dataPoint = new SurveyedLocale();
+            dataPoint.setIdentifier(String.valueOf(i));
+            dataPoint.setSurveyGroupId(surveyId);
+            dataPoint.setDisplayName("dataPoint: " + i);
+            datapoints.add(dataPoint);
+        }
+        return new ArrayList<>(dpDao.save(datapoints));
+    }
+
+    private Set<Long> getEntityIds(List<? extends BaseDomain> entities) {
+        Set<Long> entityIds = new HashSet<>();
+        for (BaseDomain o : entities) {
+            entityIds.add(o.getKey().getId());
+        }
+        return entityIds;
+    }
+
+    @Test
+    public void datastoreReadWriteTest() {
+        final Long assignmentId = 1L;
+        final Long deviceId = 1L;
+        final List<Long> dataPointIds = Collections.singletonList(0L);
+        final Long surveyId = 1L;
+
+        createDataPointAssignment(assignmentId, deviceId, dataPointIds, surveyId);
+
+        final DataPointAssignmentDao dataPointAssignmentDao = new DataPointAssignmentDao();
+        final List<DataPointAssignment> dataPointAssignments = dataPointAssignmentDao.listByDeviceAndSurvey(deviceId, surveyId);
+        assertFalse(dataPointAssignments.isEmpty());
+
+        final DataPointAssignment assignment = dataPointAssignments.get(0);
+        assertEquals(dataPointIds, assignment.getDataPointIds());
+        assertEquals(deviceId, assignment.getDeviceId());
+        assertEquals(surveyId, assignment.getSurveyId());
+
+
+        final List<Long> deviceIds = Arrays.asList(1L, 2L);
+        final List<Long> formIds = Arrays.asList(100L, 200L);
+
+        createAssignment(surveyId, deviceIds, formIds);
+
+        final SurveyAssignmentDao saDao = new SurveyAssignmentDao();
+        final List<SurveyAssignment> surveyAssignments = saDao.listAllContainingDevice(2L);
+        assertFalse(surveyAssignments.isEmpty());
+
+        final SurveyAssignment sa = surveyAssignments.get(0);
+        assertEquals(surveyId, sa.getSurveyId());
+        assertTrue(sa.getDeviceIds().contains(2L));
+        assertEquals(formIds, sa.getFormIds());
+    }
+
+    @Test
+    public void someDataPointsTest() {
+
+        final Long surveyId = 100L;
+        final List<SurveyedLocale> dataPoints = createDataPoints(surveyId, 10);
+
+        final SurveyedLocale dataPoint = dataPoints.get(7);
+        assertEquals("7", dataPoint.getIdentifier());
+
+        final List<Long> selectedDataPointIds = Arrays.asList(dataPoints.get(5).getKey().getId(), dataPoints.get(6).getKey().getId());
+        final Long assignmentId = 20L;
+        final Long deviceId = 1L;
+
+        createDataPointAssignment(assignmentId, deviceId, selectedDataPointIds, surveyId);
+
+        final DataPointServlet servlet = new DataPointServlet();
```

Comment:
This use of `servlet` was really unexpected. I was expecting all the tests to be testing at the same layer, so either the DAO layer or the servlet layer.